xarray/core/rolling_exp.py

```diff
@@ -11,15 +11,17 @@
 from xarray.core.types import T_DataWithCoords
 
 
-def _get_alpha(com=None, span=None, halflife=None, alpha=None):
+def _get_alpha(
+    com: float | None, span: float | None, halflife: float | None, alpha: float | None
+) -> float:
     # pandas defines in terms of com (converting to alpha in the algo)
     # so use its function to get a com and then convert to alpha
 
     com = _get_center_of_mass(com, span, halflife, alpha)
     return 1 / (1 + com)
 
 
-def move_exp_nanmean(array, *, axis, alpha):
+def move_exp_nanmean(array: np.ndarray, *, axis: int, alpha: float) -> np.ndarray:
```

Comment:
Is it really true that array can only be np.ndarray? cupy arrays are not allowed?

Reply:
I hadn't thought of that — I think it is possible: https://docs.cupy.dev/en/stable/user_guide/interoperability.html#numba

Do you know what the correct item to use here is? `ArrayLike`?

Reply:
In #8075 we are testing `T_DuckArray = typing.TypeVar("T_DuckArray", bound=typing.Any)` for `self.data`, which should imply "an array that follows the array api".

Nothing is in main at the moment though, probably because it would open so many cans of worms. :D

 I think the important thing here though is that we just have to put a `T_DuckArray` in `types.py` and start to use it anywhere `Variable/DataArray.data` is used.

Reply:
OK — what should do we do for this PR? This is just trying to start a pattern of allowing some strict modes — I hadn't really intended it to be breaking new ground on array types... :)

I could replace with `Any` if we would rather have something vague that slightly wrong...

Reply:
In my opinion, there's no point in strict mode if we're just going to use Any. 
Add T_DuckArray in types.py and use it here, hopefully mypy won't complain too much.

Reply:
If #8075 merges soon, I'm happy to use that — otherwise I really did just open this to start an template for expanding strict mode, I would strongly prefer to keep this PR focused & generally allow PRs to be small marginal improvements rather than midas-touching lots of code.

(and while we should indeed attempt to not use `Any`, it's still better than nothing, because a function with some annotations still works to check types internally)

Reply:
Added `T_DuckArray`!